### PR TITLE
Further CLI Improvements

### PR DESCRIPTION
**Pull request by fzdarsky**
Uses a common structure for options and how they are completed, validated, and run.
Cleans up description of commands and flags.
Adds support for JSON as output format for get command.

Depends on #136

### CODE REVIEW

**Review by mangelajo — APPROVED** [review]
Looks great frank!

File: internal/cli/approve.go [hunk 1]
```diff
@@ -11,36 +11,54 @@ import (
...
+				return err
+			}
+			return o.Run(cmd.Context(), args)
 		},
 		SilenceUsage: true,
 	}
 
-	cmd.Flags().StringArrayVarP(&approveLabels, "label", "l", []string{}, "labels to add to the device")
-	cmd.Flags().StringVarP(&approveRegion, "region", "r", "default", "region for the device")
+	cmd.Flags().StringArrayVarP(&o.ApproveLabels, "label", "l", []string{}, "Labels to add to the device, as a comma-separated list of key=value.")
```
Comment (oglok) [inline internal/cli/approve.go:37]: while Use and Short is ok to start with capital letters, I think the standard for flags is lower case. Not strict rule though.

File: internal/cli/apply.go [hunk 2]
```diff
@@ -32,25 +32,23 @@ type ApplyOptions struct {
...
 	cmd := &cobra.Command{
-		Use:                   "apply -f FILENAME",
-		DisableFlagsInUseLine: true,
-		Short:                 "apply a configuration to a resource by file name or stdin",
+		Use:   "apply -f FILENAME",
+		Short: "Apply a configuration to a resource by file name or stdin.",
 		RunE: func(cmd *cobra.Command, args []string) error {
-			if !cmd.Flags().Lookup("filename").Changed {
-				return fmt.Errorf("must specify -f FILENAME")
+			if err := o.Complete(cmd, args); err != nil {
```
[the remark this answers is not in the document]
Comment (hexfusion) [reply internal/cli/apply.go:39]: I believe the idea is to reuse the pattern on every command. 

<complete>
<verify>
<run>

Just for now some methods will not have any work but probably will in the future?

File: internal/cli/apply.go [hunk 3]
```diff
@@ -59,12 +57,86 @@ func NewCmdApply() *cobra.Command {
...
+		default:
+			expandedFilenames, err := expandIfFilePattern(filename)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			for _, filename := range expandedFilenames {
+				_, err := os.Stat(filename)
+				if os.IsNotExist(err) {
+					errs = append(errs, fmt.Errorf("the path %q does not exist", filename))
```
Comment (hexfusion) [inline internal/cli/apply.go:100]: My only question is should we fail fast here?

**Review by hexfusion — APPROVED** [review]

File: internal/cli/apply.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fzdarsky) [reply internal/cli/apply.go:39]: Yes, it's to have a common structure and a place for adding code in the future when we need to fill in options prior to validation. Structure is that also used in K8s.

File: internal/cli/approve.go (hunk 1 above)
Comment (fzdarsky) [reply internal/cli/approve.go:37]: Do you a pointer to best-practices or style guides? I'm ok either way, but this follows what kubectl standardises on.

File: internal/cli/apply.go (hunk 3 above)
Comment (fzdarsky) [reply internal/cli/apply.go:100]: IDK. Maybe? I only moved this method up without changing it. Q is whether we want to greedily apply as many resources as possible or to stop on the first error. What's K8s' behavior?

Comment (hexfusion) [reply internal/cli/apply.go:100]: I think this is considered a preflight check, kube bails
```
oc apply -f test.yaml,test.broke
error: the path "test.broke" does not exist
```

Comment (hexfusion) [reply internal/cli/apply.go:100]:
```
oc apply -f test.yaml,test.broke,test.broke2
the path "test.broke" does not exist
the path "test.broke2" does not exist
```

**Comment by mangelajo** [comment]
I say we merge this one, and continue improving.

**Review by hexfusion — APPROVED** [review]

**Pull request merged**